Readme.md
# gptaddin <a href="https://crumplab.com/gptaddin"><img src="man/figures/logo.png" align="right" height="120" /></a>

<!-- badges: start -->
  [![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
  <!-- badges: end -->

The purpose of this package is to create RStudio addins for interfacing with the [OpenAI API](https://openai.com/) through the [openai R library](https://github.com/irudnyts/openai). 

The add-ins are intended to provide writing assistant functions, such as taking selected text in the RStudio editor and having a GPT model return edits.

This is an experimental project undergoing a short spurt of development for my personal use.

I'm not encouraging anyone to install this package, but it could serve as a useful example for other people looking to customize their own writing assistant in RStudio.

I blogged about the process of making and using this package over a few posts:

<https://crumplab.com/blog/770_rstudio_gpt/>

<https://crumplab.com/blog/668_gpt_adventure/>

<https://crumplab.com/blog/666_gpt_grammar/>

See the example page for some more explanation and ongoing dev notes.

<https://crumplab.com/gptaddin/articles/examples.html>

### Usage notes

I don't currently have plans to make this package ready for a larger audience to install and use without any hassle. There are several privacy concerns to consider when sending data to OpenAI, and the behavior of LLMs can be unstable. In my opinion, interacting with these tools requires a great deal of human supervision. For example, while using the functions in this package myself, I rarely get results that are useful straight away. The reason I'm creating these functions is to personally understand what I am sending and receiving, as well as to have programmatic control over my actions. For these reasons, as I mentioned earlier, I'm not suggesting that anyone should install and use this package.

In case this rough scratchpad of a package is useful for others, here's a quick rundown.

It is possible to install this package from github, **although, I'm not recommending this option**:

```
#install.packages("devtools")
devtools::install_github("CrumpLab/gptaddin")
```

I imagine it would be more useful to fork, clone, or otherwise download the repo, install it locally, and then modify it as needed for your use case. 

1. Download the repo to your computer
2. Open the .Rproj file in Rstudio
3. Install the package from the build tab.

I'm following the basic skeleton for making R packages using pkgdown. See the [R packages book](https://r-pkgs.org) for more info.

#### Addins

Once this package is installed, several addins should become available in Rstudio. See the [examples page](https://crumplab.com/gptaddin/articles/examples.html) for some tips on usage. 

I wrote these addins to explore convenient ways to use the OpenAI API while using RStudio. They do things like send highlighted text in the RStudio editor to an LLM, and then print results to the console, or to the editor window. 

**This process only works if you have an account with OpenAI, and you have created an API key**.

1.  Create a secret key in openai, copy it and put it somewhere safe. Don't share it.

2.  The `openai` R library looks for an environment variable called `OPENAI_API_KEY`, which can be set a [number of ways](https://irudnyts.github.io/openai/).

My strategy was to add a key-value pair directly to the `.Renviron` file at the user level.

```{r}
#| eval: false

# open the .Renviron file
usethis::edit_r_environ()
```

Then edit the file to add the new key-value pair, and restart R.

```{r}
#| eval: false

# add new environment variable to above and save
OPENAI_API_KEY=XXXXXXXXXXXXX
```

After you restart R, you can check if R recognizes the new environment variable:

```{r}
#| eval: false
Sys.getenv("OPENAI_API_KEY")
```

**The functions in this package assume the above, and would need to be modified to handle alternate ways of sending the API key.**

#### Writing addins

Some of the addins in this package may do things you want them to. If not, I suggest writing your own to achieve your desired workflow.

Documentation for writing custom addins is [here](https://RStudio.github.io/RStudioaddins/). RStudio addins are available from a dropdown menu and/or triggered by hot-key macros.

The steps for writing addins are:

1.  Write functions as you normally would do in the R folder. You can see the functions I wrote in `R/addins.R`
3.  [Register the functions as addins](https://rstudio.github.io/rstudioaddins/#registering-addins) using a .dcf file placed in `inst/rstudio/addins.dcf`
4.  Install the R package from the build tab anytime you make changes
5.  The addins should now be available for use in RStudio

#### Shiny app for grammar checking

The newest piece to this package is a Shiny App for grammar checking. I'm starting to get decent results with this approach. The functions for the shiny app are in `R/grammar_shiny.R`.

```
gptaddin::run_grammar_checker("yourfile.qmd")
```

### Other examples of using GPT in RStudio

- [gptstudio](https://github.com/MichelNivard/gptstudio)









NEWS.md
# gptaddin 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Added custom prompt option to Shiny app
* Added gpt-4 option in Shiny app
* Allow user to send custom text by editing the text window in Shiny app.


R/addins.R

#' Get suggested line edits
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a gpt-3.5-turbo chat model. The system prompt is:
#'
#' You are an editorial writing assistant. Your task is to edit the content you receive for improved clarity, flow, and reduced word count.
#'
#' The model receives the highlighted text, generates a response following the prompt, and returns the new suggested text to the console.
#'
#' @param model_type string, default "gpt-3.5-turbo"
#'
#' @return string to console
#' @export
#'
line_editor <- function(model_type = "gpt-3.5-turbo"){


  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are an editorial writing assistant. Your task is to edit the content you receive for improved clarity, flow, and reduced word count. You should respect the original style of the input text and use an active voice."
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}

#' Get suggested line edits
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu. Allows a selection of different editing prompts from the console before sending.
#'
#' This function sends the selected text to a gpt-3.5-turbo chat model. The system prompt is:
#'
#' You are an editorial writing assistant. Your task is to edit the content you receive for improved clarity, flow, and reduced word count.
#'
#' The model receives the highlighted text, generates a response following the prompt, and returns the new suggested text to the console.
#'
#' @param model_type string, default "gpt-3.5-turbo"
#'
#' @return string to console
#' @export
#'
gpt_edit <- function(model_type = "gpt-3.5-turbo"){

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  user_choice <- menu(
    choices = c(
      "Basic spelling and grammar",
      "Flow, reduced word count",
      "Compress text",
      "Write custom prompt: "
    ),
    title = "Choose editing prompt"
  )

  if( user_choice == 1){
    system_content <- "You are an editorial writing assistant. Edit the text for spelling and grammar. Don't change the meaning or the words."
  }

  if( user_choice == 2){
    system_content <- "You are an editorial writing assistant. Your task is to edit the content you receive for improved clarity, flow, and reduced word count. You should respect the original style of the input text and use an active voice."
  }

  if( user_choice == 3){
    system_content <- "You are a text compressor. Compress the meaning of the text into as few words as possible, make it so that you can understand the text even if it is not readable to a human."
  }

  if( user_choice == 4){
    system_content <- readline(prompt = "Enter your prompt")
  }

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = system_content
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}

#' Get suggested line edits with markdown flagged changes
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a gpt-3.5-turbo chat model. The system prompt is:
#'
#' You are an editorial writing assistant. Your task is to edit the content you receive for improved clarity, flow, and reduced word count. You should respect the original style of the input text and use an active voice. Please return the edited text using markdown syntax to highlight all the changed wording. Specifically, put bold all of the changed words or phrases that are different from the original text.
#'
#' The model receives the highlighted text, generates a response following the prompt, and returns the new suggested text to the console.
#'
#' @param model_type string, default "gpt-3.5-turbo"
#'
#' @return string to console
#' @export
#'
line_editor_markdown <- function(model_type = "gpt-3.5-turbo"){


  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are an editorial writing assistant. Your task is to edit the content you receive for improved clarity, flow, and reduced word count. You should respect the original style of the input text and use an active voice. Please return the edited text using markdown syntax to bold any of the suggested changes."
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}



#' Convert point form to longform prose
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight point form text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a gpt-3.5-turbo chat model. The system prompt is:
#'
#' You are a writing assistant. Your task is to take point form notes and convert them into longform prose. Please write in a straightforward and clear manner. The user prompt will contain point form notes.
#'
#' The model receives the highlighted text, generates a response following the prompt, and returns the new suggested text to the console.
#'
#' @param model_type string, default "gpt-3.5-turbo"
#'
#' @return string to console
#' @export
#'
points_to_prose <- function(model_type = "gpt-3.5-turbo"){


  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are a writing assistant. Your task is to take point form notes and convert them into longform prose. Please write in a straightforward and clear manner. The user prompt will contain point form notes."
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}


#' summarize text into a series of points
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a gpt-3.5-turbo chat model. The system prompt is:
#'
#' You are a writing assistant. Your task is to summarize input text into clear pointform notes.
#'
#' The model receives the highlighted text, generates a response following the prompt, and returns the new suggested text to the console.
#'
#' @param model_type string, default "gpt-3.5-turbo"
#'
#' @return string to console
#' @export
summarize_main_points <- function(model_type = "gpt-3.5-turbo"){

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are a writing assistant. Your task is to summarize input text into clear pointform notes."
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}

#' Send highlighted text as completion prompt to ada
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to an ada model
#'
#' The model receives the highlighted text, generates a completion prompt, and returns the generated text to the console.
#'
#' @return string to console
#' @export
completion_ada <- function(){

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")
  message("Your prompt was")
  message(selected_text)

  # run the api call to openai
  gpt <- openai::create_completion(
    model = "text-ada-001",
    prompt = selected_text$value,
    max_tokens = 256,

  )

  # print the results to the console
  print(cat(gpt$choices$text))

}

#' Send highlighted text as completion prompt to davinci
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a davinci model
#'
#' The model receives the highlighted text, generates a completion prompt, and returns the generated text to the console.
#'
#' @return string to console
#' @export
completion_davinci <- function(){

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")
  message("Your prompt was")
  message(selected_text)

  # run the api call to openai
  gpt <- openai::create_completion(
    model = "text-davinci-003",
    prompt = selected_text$value,
    max_tokens = 256
  )

  # print the results to the console
  print(cat(gpt$choices$text))

}


#' Send highlighted text as a completion prompt to gpt-4
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a gpt-4 model
#'
#' The model receives the highlighted text, generates a completion prompt, and returns the generated text to the console.
#'
#' @param model_type string, default "gpt-4"
#'
#' @return string to console
#' @export
completion_chat <- function(model_type = "gpt-4"){

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are a helpful assistant."
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}

#' Send highlighted text to gpt-4 output to doc
#'
#' This is an RStudio addin function that should become available after the package is installed.
#'
#' Highlight text in an editor window while using Rstudio, then select the addin from the drop-down menu.
#'
#' This function sends the selected text to a gpt-4 model and returns the output to the document, not the console.
#'
#' The selected text is replaced with the original text, a new line, and the response from the model.
#'
#' @param model_type string, default "gpt-4"
#'
#' @return string to console
#' @export
chat_to_doc <- function(model_type = "gpt-4"){

  # get document id and selection range
  active_context <- rstudioapi::getActiveDocumentContext()
  #print(active_context)

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  #run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are a helpful assistant."
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # insert original selection plus response back into document
  new_text <- gpt$choices$message.content
  insert_text <- paste(selected_text,"\n", new_text, sep="\n")

  rstudioapi::insertText(location = active_context$selection[[1]]$range,
                         text = insert_text,
                         id = active_context$id)
}

#' Start a continuous chat with message history
#'
#' Run this in the console
#'
#' Very basic right now, no error checking. Chat history is preserved in my_chat which is returned to the global environment. Running this function initializes my_chat, and deletes anything that was in there.
#'
#' @param model_type string, default "gpt-4"
#'
#' @return list called my_chat to the global environment
#'
#' @export
start_chat <- function(model_type = "gpt-4"){

  # make list to save results
  my_chat <<- list()

  # initial prompt
  my_chat$message <<- list(
    list("role" = "system",
         "content" = "You are a helpful assistant.")
  )

  # send prompt
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = my_chat$message
  )

  #store results
  my_chat$history[[1]] <<- gpt
  my_chat$message[[length(my_chat$message)+1]] <<- list(
    "role" = gpt$choices$message.role,
    "content" = gpt$choices$message.content
  )

  # print response to console
  print(cat(gpt$choices$message.content))

}

#' Continue a chat with message history
#'
#' To be used in combination with start_chat
#'
#' Select any text in the editor then run this addin.
#'
#' The selected text is combined with the existing message history and sent as a prompt to OpenAI.
#'
#' @param model_type string, default "gpt-4"
#'
#' @return string to the editor window
#'
#' @export
continue_chat <- function(model_type="gpt-4"){

  # get document id and selection range
  active_context <- rstudioapi::getActiveDocumentContext()
  #print(active_context)

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  my_chat$message[[length(my_chat$message)+1]] <<- list(
    "role" = "user",
    "content" = selected_text$value
  )

  # send prompt
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = my_chat$message
  )

  #store results
  my_chat$history[[length(my_chat$history)+1]] <<- gpt
  my_chat$message[[length(my_chat$message)+1]] <<- list(
    "role" = gpt$choices$message.role,
    "content" = gpt$choices$message.content
  )

  print(paste("Prompt tokens used: ",my_chat$history[[length(my_chat$history)]]$usage$prompt_tokens))

  new_text <- gpt$choices$message.content
  insert_text <- paste(selected_text,"\n", new_text, sep="\n")

  rstudioapi::insertText(location = active_context$selection[[1]]$range,
                         text = insert_text,
                         id = active_context$id)
}

#' Generates documentation in Roxygen syntax for an R function
#'
#' This addin is registered and accessible through RStudio. The input to the function is any highlighted text in the RStudio editor.
#'
#' @param model_type string, default "gpt-4"
#'
#' @return the documentation is printed to the console
#'
#' @examples
#' #write_function_doc()
#'
#' @export
write_function_doc <- function(model_type="gpt-4"){

  # get selected text
  selected_text <- rstudioapi::selectionGet()

  message("Submitting query...")

  # run the api call to openai
  gpt <- openai::create_chat_completion(
    model = model_type,
    messages = list(
      list(
        "role" = "system",
        "content" = "You are an expert R programmer capable of writing clear function documentation in the Roxygen syntax. You will receive an R function, and return the function documentation in Roxygen syntax."
      ),
      list(
        "role" = "assistant",
        "content" = "Ok, what is the function?"
      ),
      list(
        "role" = "user",
        "content" = selected_text$value
      )
    )
  )

  # print the results to the console
  print(cat(gpt$choices$message.content))
}




R/grammar_shiny.R
# Functions for the Shiny app

#' Edit text for spelling, grammar, word count, clarity using GPT-3
#'
#' This function uses GPT-3 to assist in editing text for spelling, grammar, word count, and clarity. It is a companion function for the run grammar checker Shiny app. The function takes in the following parameters:
#'
#' @param text_to_edit: The original text that needs editing.
#' @param user_choice: A character value indicating the type of editing required. Possible values include "spelling_grammar", "word_count", or "clarity".
#' @param model_type: A character value indicating the GPT-3 model to use, these are supplied by the shiny app.
#'
#' @return Returns the edited text as a string.
#'
#' @export
check_grammar_gpt <- function(text_to_edit, user_choice, model_type, custom_prompt){

  # get selected text
  selected_text <- text_to_edit

  if( user_choice == "spelling_grammar"){
    system_content <- "You are an editorial writing assistant. Edit the text for spelling and grammar. Don't change the meaning of the words."
  }

  if( user_choice == "word_count"){
    system_content <- "You are an editorial writing assistant. Edit the text to reduce word count without changing the meaning."
  }

  if( user_choice == "clarity"){
    system_content <- "You are an editorial writing assistant. Edit the text to improve clarity and flow."
  }

  if( user_choice == "custom"){
    system_content <- custom_prompt
  }



  if( model_type %in% c("text-davinci-003","text-curie-001","text-babbage-001","text-ada-001")){
    gpt <- openai::create_completion(
      model = model_type,
      prompt = paste(system_content,"\n",selected_text),
      max_tokens = 1000
    )

    model_response <- gpt$choices$text
  }

  if( model_type %in% c("gpt-3.5-turbo","gpt-4")){
    gpt <- openai::create_chat_completion(
      model = model_type,
      messages = list(
        list(
          "role" = "system",
          "content" = system_content
        ),
        list(
          "role" = "assistant",
          "content" = "Ok, please send the text."
        ),
        list(
          "role" = "user",
          "content" = selected_text
        )
      )
    )

    model_response <- gpt$choices$message.content
  }

  #print(gpt) error-checking
  my_string <- gsub("\n", "", model_response)

  return(my_string)
}

#' Compare Two Texts
#'
#' This function compares two texts and returns whether they are identical or a diff object from diffobj to visualize the differences in HTML. Used in conjunction with the run grammar checker Shiny app.
#'
#' @param original_text A character string representing the original text.
#' @param modified_text A character string representing the modified text.
#'
#' @return Either "identical" if the two texts are identical, or a diff object showing the differences between the two texts.
#'
#' @export
compare_text <- function(original_text, modified_text) {
  if (identical(original_text, modified_text)) {
    return("identical")
  } else {
      as.character(
        diffobj::diffChr(original_text, modified_text, format="html",
                           style=list(html.output="diff.w.style"))
      )
  }
}

# These functions were inspired by, and rewritten from https://github.com/jasdumas/gramr

#' Parse RMarkdown documents into sentences or paragraphs
#' @param file a character vector of the file or path
#' @examples
#' # don't run
#' # parse_rmd()
#' @export
parse_rmd <- function(file,token_level="paragraphs"){

  # Read in the R Markdown file
  rmd_file <- readLines(file)

  if(token_level == "sentences"){
    # Parse the Rmd file into individual sentences
    sentences <- unlist(tokenizers::tokenize_sentences(rmd_file))
  }

  if(token_level == "paragraphs"){
    # Parse the Rmd file into individual sentences
    sentences <- unlist(tokenizers::tokenize_paragraphs(rmd_file))
  }

  sentences

}


#' Start grammar checker shiny application
#'
#' This shiny app allows a user to manually check each sentence or paragraph in a file using OpenAI LLMs.
#'
#' The text file is split into sentences or paragraphs. The user manually goes through each and chooses whether or not to send the current text for suggested feedback.
#'
#' The prompts used for each checking option are defined in check_grammar_gpt.
#'
#' Any suggested changes returned by the model are displayed with differences between the original and modified text highlighted in the window. The returned text often has subtle changes that may be difficult to spot, hence the visualization approach.
#'
#' The suggested edits may occasionally fail to follow the instructions from the prompt. Some of the models produce very low-quality results, like ada, and babbage.
#'
#' @param path file to check
#' @param choose_token_level string "sentences" or "paragraphs", defaults to paragraphs. The document is parsed into individual sentences or paragraphs.
#' @return a shiny app is launched
#' @import shiny
#' @examples
#' # don't run during tests
#' # gptaddin::run_grammar_checker("example.rmd")
#' @export
run_grammar_checker <- function(path, choose_token_level = "paragraphs"){
  sentences <- parse_rmd(path, choose_token_level)
  counter <- 1

  ui <- fluidPage(
    # Application title
    titlePanel(
      "GPT Grammar checker"
    ),
    sidebarLayout(
      sidebarPanel(
        selectInput(inputId = 'options',
                           label = "Choose prompt",
                           choices = list("Basic Spelling and grammar" = "spelling_grammar",
                                            "Reduced word count" = "word_count",
                                            "Improved clarity" = "clarity",
                                            "Custom" = "custom"),
                            multiple = FALSE
        ),
        selectInput(inputId = 'model',
                    label = "Choose model",
                    choices = list("gpt-3.5-turbo $.002" = "gpt-3.5-turbo",
                                   "gpt-4 $.03" = "gpt-4",
                                   "Davinci $.03" = "text-davinci-003",
                                   "Curie $.002" = "text-curie-001",
                                   "Babbage $.0005" = "text-babbage-001",
                                   "Ada $.0004" = "text-ada-001"),
                    multiple = FALSE
        ),
        sliderInput(inputId = "doc_location",
                    "Scroll Document",
                    min = 1,
                    max = length(sentences),
                    value = 1),
        actionButton("nextS", "Next"),
        actionButton("submit","Submit"),
        actionButton("backS", "Back"),
        actionButton("exit", "Finish"),
        textAreaInput(inputId  = 'system_text',
                      label  = 'System Role',
                      value  = "You are a blah blah blah, that will do something to the text.",
                      height = 100,
                      resize = "both"),
        textAreaInput(inputId  = 'assistant_text',
                      label  = 'Assistant Role',
                      value  = "Ok, please send the text",
                      height = 100,
                      resize = "both")
      ),
      mainPanel(
        textAreaInput(inputId  = 'text',
                      label  = 'Text to Check',
                      value  = sentences[1],
                      height = 150,
                      width = 500,
                      resize = "both"),
        h3("Suggestion"),
        htmlOutput(outputId = "suggested_text"),
        actionButton("copy", "Copy to clipboard")
      )
    )
  )

  server <- function(input, output, session) {

    ## Initialize Reactive variables ####
    values <- reactiveValues()

    # go to next sentence
    observeEvent(input$nextS, {
      if( counter < length(sentences)){
        counter <<- counter + 1
        updateTextInput(session, "text", value = sentences[counter])
      }
    })

    # go to previous sentence
    observeEvent(input$backS, {
      if( counter > 1){
        counter <<- counter - 1
        updateTextInput(session, "text", value = sentences[counter])
      }
    })

    # scroll through document
    observeEvent(input$doc_location, {
        counter <<- input$doc_location
        updateTextInput(session, "text", value = sentences[counter])
    })

    # submit sentence for checking to OpenAI API
    observeEvent(input$submit, {
      #print(sentences[counter])
      #print(input$options)
      #print(input$model)
      #print(input$text)

      #pass text to API and get result
      #if user has modified the text, the text in the box will be sent
      values$suggestion <- check_grammar_gpt(input$text,
                                      input$options,
                                      input$model,
                                      input$system_text)

      # find differences between original and modified text
      diff_1 <- compare_text(input$text,values$suggestion)

      # display differences
      output$suggested_text <- renderUI({
        HTML(diff_1)
      })

    })

    # copy to clipboard
    observeEvent(input$copy, {
      clipr::write_clip(values$suggestion)
    })

    # exit the app
    observeEvent(input$exit, {
      stopApp()
    })
  }

  # Run the application
  shinyApp(ui = ui, server = server)
}


